File: core/src/com/ultreon/craft/registry/RegistrySupplier.java
package com.ultreon.craft.registry;

import com.ultreon.craft.resources.Identifier;

import java.util.function.Supplier;

public class RegistrySupplier<T> implements Supplier<T> {
    private final Registry<? super T> registry;
    private final Identifier id;

    public RegistrySupplier(Registry<? super T> registry, Identifier id) {
        this.registry = registry;
        this.id = id;
    }

    @Override
    @SuppressWarnings("unchecked")
    public T get() {
        T t = (T) registry.get(id);
        if (t == null) {
            throw new IllegalArgumentException("Registry entry not present: " + id);
        }
        return t;
    }
}

File: src/jvmMain/kotlin/com/ultreon/craft/registry/RegistrySupplier.kt
package com.ultreon.craft.registry

import com.ultreon.craft.resources.*
import com.ultreon.craft.util.*

class RegistrySupplier<T : Any>(
    private val registry: Registry<in T>,
    private val id: Identifier
) {
    @Suppress("UNCHECKED_CAST")
    operator fun invoke(): T {
        if (id !in registry)
            throw IllegalArgumentException("Object not registered: $id")
        return registry[id] as T
    }
}
